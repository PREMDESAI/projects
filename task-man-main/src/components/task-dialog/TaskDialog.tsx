import {
  Button,
  Dialog,
  Flex,
  Select,
  Text,
  TextArea,
  TextField,
} from '@radix-ui/themes';
import React from 'react';
import { useNavigate, useParams } from 'react-router-dom';

import { getStatusText } from '@/helpers';
import { TaskStatus, useTasks } from '@/hooks/use-tasks';

export const TaskDialog = () => {
  const { tasksView, statuses, addTask, updateTask } = useTasks();
  const navigate = useNavigate();
  const { status, id } = useParams<{ status?: TaskStatus; id?: string }>();
  const selectedTask = React.useMemo(
    () => tasksView.find((task) => task.id === id),
    [id, tasksView]
  );

  const isUpdateMode = Boolean(selectedTask);

  const [title, setTitle] = React.useState(selectedTask?.title);
  const [description, setDescription] = React.useState(
    selectedTask?.description || ''
  );
  const [taskStatus, setTaskStatus] = React.useState<TaskStatus | undefined>(
    status || selectedTask?.status
  );

  const [isOpen, setIsOpen] = React.useState(true);

  if (status) {
    setTimeout(() => navigate('/new', { replace: true }));
  }

  const isFormValid = Boolean(title);

  const onSubmit = async () => {
    if (isFormValid) {
      if (isUpdateMode) {
        await updateTask(selectedTask!.id, {
          title: title!,
          status: taskStatus!,
          description,
        });
      } else {
        await addTask({
          title: title!,
          status: taskStatus!,
          description,
        });
      }

      setIsOpen(false);
    }
  };

  return (
    <Dialog.Root
      open={isOpen}
      onOpenChange={(open) => !open && navigate('/', { replace: true })}
    >
      <Dialog.Content maxWidth='450px'>
        <Dialog.Title>{isUpdateMode ? 'Update Task' : 'New Task'}</Dialog.Title>
        <Dialog.Description size='2' mb='4'>
          Add a new task to your list
        </Dialog.Description>

        <Flex direction='column' gap='3'>
          <label>
            <Text as='div' size='2' mb='1' weight='bold'>
              Title
            </Text>
            <TextField.Root
              placeholder='What would you like to achieve ?'
              required
              defaultValue={title}
              onChange={(event) => setTitle(event.target.value)}
              aria-label='Task Title'
            />
          </label>
          <label>
            <Text as='div' size='2' mb='1'>
              Description
            </Text>
            <TextArea
              placeholder='Elaborate a little...'
              onChange={(event) => setDescription(event.target.value)}
              defaultValue={description}
              aria-label='Task Description'
            />
          </label>
          <label>
            <Text as='div' size='2' mb='1'>
              Status
            </Text>
            <Select.Root
              defaultValue={taskStatus || statuses[0]}
              onValueChange={(value) => setTaskStatus(value as TaskStatus)}
            >
              <Select.Trigger />
              <Select.Content>
                {statuses.map((status) => (
                  <Select.Item
                    value={status}
                    key={`task-dialog-status-${status}`}
                  >
                    {getStatusText(status)}
                  </Select.Item>
                ))}
              </Select.Content>
            </Select.Root>
          </label>
        </Flex>

        <Flex gap='3' mt='4' justify='end'>
          <Dialog.Close>
            <Button variant='soft' color='gray'>
              Cancel
            </Button>
          </Dialog.Close>
          <Dialog.Close>
            <Button disabled={!isFormValid} onClick={onSubmit}>
              Save
            </Button>
          </Dialog.Close>
        </Flex>
      </Dialog.Content>
    </Dialog.Root>
  );
};
