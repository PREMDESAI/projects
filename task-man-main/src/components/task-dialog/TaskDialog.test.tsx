import { render, screen, waitFor } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { describe, it } from 'vitest';
import userEvent from '@testing-library/user-event';

import { TaskDialog } from './TaskDialog';

describe('TaskDialog', () => {
  it('renders', async () => {
    render(
      <MemoryRouter>
        <TaskDialog />
      </MemoryRouter>
    );

    const dialog = screen.queryByRole('dialog');
    await waitFor(() => expect(dialog).toBeVisible());
    expect(dialog).toMatchSnapshot();
  });

  it('does not let user submit the form unless title is provided', async () => {
    render(
      <MemoryRouter>
        <TaskDialog />
      </MemoryRouter>
    );

    const dialog = screen.queryByRole('dialog');
    await waitFor(() => expect(dialog).toBeVisible());

    expect(screen.getByRole('button', { name: 'Save' })).toBeDisabled();

    await userEvent.type(
      screen.getByRole('textbox', { name: 'Task Title' }),
      'New Task'
    );

    expect(screen.getByRole('button', { name: 'Save' })).toBeEnabled();
  });
});
