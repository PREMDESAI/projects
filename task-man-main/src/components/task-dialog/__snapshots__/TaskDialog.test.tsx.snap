// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TaskDialog > renders 1`] = `
<div
  aria-describedby="radix-:r2:"
  aria-labelledby="radix-:r1:"
  class="rt-BaseDialogContent rt-DialogContent rt-r-size-3 rt-r-max-w"
  data-state="open"
  id="radix-:r0:"
  role="dialog"
  style="--max-width: 450px; pointer-events: auto;"
  tabindex="-1"
>
  <h1
    class="rt-Heading rt-r-size-5 rt-r-lt-start rt-r-mb-3"
    id="radix-:r1:"
  >
    New Task
  </h1>
  <p
    class="rt-Text rt-r-size-2 rt-r-mb-4"
    id="radix-:r2:"
  >
    Add a new task to your list
  </p>
  <div
    class="rt-Flex rt-r-fd-column rt-r-gap-3"
  >
    <label>
      <div
        class="rt-Text rt-r-size-2 rt-r-weight-bold rt-r-mb-1"
      >
        Title
      </div>
      <div
        class="rt-TextFieldRoot rt-r-size-2 rt-variant-surface"
      >
        <input
          aria-label="Task Title"
          class="rt-reset rt-TextFieldInput"
          placeholder="What would you like to achieve ?"
          required=""
          spellcheck="false"
          value=""
        />
      </div>
    </label>
    <label>
      <div
        class="rt-Text rt-r-size-2 rt-r-mb-1"
      >
        Description
      </div>
      <div
        class="rt-TextAreaRoot rt-r-size-2 rt-variant-surface"
      >
        <textarea
          aria-label="Task Description"
          class="rt-reset rt-TextAreaInput"
          placeholder="Elaborate a little..."
        />
      </div>
    </label>
    <label>
      <div
        class="rt-Text rt-r-size-2 rt-r-mb-1"
      >
        Status
      </div>
      <button
        aria-autocomplete="none"
        aria-controls="radix-:r3:"
        aria-expanded="false"
        class="rt-reset rt-SelectTrigger rt-r-size-2 rt-variant-surface"
        data-placeholder=""
        data-state="closed"
        dir="ltr"
        role="combobox"
        type="button"
      >
        <span
          class="rt-SelectTriggerInner"
        >
          <span
            style="pointer-events: none;"
          />
        </span>
        <svg
          aria-hidden="true"
          class="rt-SelectIcon"
          fill="currentcolor"
          height="9"
          viewBox="0 0 9 9"
          width="9"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0.135232 3.15803C0.324102 2.95657 0.640521 2.94637 0.841971 3.13523L4.5 6.56464L8.158 3.13523C8.3595 2.94637 8.6759 2.95657 8.8648 3.15803C9.0536 3.35949 9.0434 3.67591 8.842 3.86477L4.84197 7.6148C4.64964 7.7951 4.35036 7.7951 4.15803 7.6148L0.158031 3.86477C-0.0434285 3.67591 -0.0536285 3.35949 0.135232 3.15803Z"
          />
        </svg>
      </button>
    </label>
  </div>
  <div
    class="rt-Flex rt-r-jc-end rt-r-gap-3 rt-r-mt-4"
  >
    <button
      class="rt-reset rt-BaseButton rt-r-size-2 rt-variant-soft rt-Button"
      data-accent-color="gray"
      type="button"
    >
      Cancel
    </button>
    <button
      class="rt-reset rt-BaseButton rt-r-size-2 rt-variant-solid rt-Button"
      data-accent-color=""
      data-disabled="true"
      disabled=""
      type="button"
    >
      Save
    </button>
  </div>
</div>
`;
